### ISSUE TITLE
Bug: wrong unicode regexp

### ISSUE THREAD

**Issue by vitaliylag** [babel/babel#6931]
Here is my code:
```js
var a = 'a',                        //usual character
    s = String.fromCharCode(55296), //start of surrogate pair
    e = String.fromCharCode(56320); //end of surrogate pair

console.log((a + e).match(/./ug).length); //must print "2" but prints "1" after babel
console.log((e + e).match(/./ug).length); //must print "2" but prints "1" after babel

```

Babel transforms it to this code:

```js
var a = 'a',
    s = String.fromCharCode(55296),
    e = String.fromCharCode(56320);

console.log((a + e).match(/(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/g).length); //must print "2" but prints "1" after babel
console.log((e + e).match(/(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/g).length); //must print "2" but prints "1" after babel
```

As you can see "." character is processed totally wrong.

Currently babel converts it to:
```js
/(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/
```

but must convert to:
```js
/(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uDC00-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]?)/
```

This not only much shorter but also does not contain such easy bugs.

I also wrote the test:
```js
'use strict';



if (typeof describe === 'undefined') {
	//Use "mocha" command to run this test (npm install -g mocha and move the file to dir with name "test")
	global.describe = (name, clb) => clb();
	global.it = (name, clb) => clb();
}



const assert = require('assert');



const a = 'a',                        //usual character
      s = String.fromCharCode(55296), //start of surrogate pair
      e = String.fromCharCode(56320); //end of surrogate pair

const native = /./ug,
      my = /(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uDC00-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]?)/g,
      babel = /(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/g;

testAll('native', native);
testAll('my', my);
testAll('babel', babel);



function testAll(name, rexp) {
	describe(name, function() {
		testOne(rexp, a + a, [a, a]);
		testOne(rexp, a + s, [a, s]);
		testOne(rexp, a + e, [a, e]);
		testOne(rexp, s + a, [s, a]);
		testOne(rexp, s + s, [s, s]);
		testOne(rexp, s + e, [s + e]);
		testOne(rexp, e + a, [e, a]);
		testOne(rexp, e + s, [e, s]);
		testOne(rexp, e + e, [e, e]);
		
		testOne(rexp, a, [a]);
		testOne(rexp, s, [s]);
		testOne(rexp, e, [e]);
		
		testOne(rexp, a + a + a, [a, a, a]);
		testOne(rexp, a + a + s, [a, a, s]);
		testOne(rexp, a + a + e, [a, a, e]);
		testOne(rexp, a + s + a, [a, s, a]);
		testOne(rexp, a + s + s, [a, s, s]);
		testOne(rexp, a + s + e, [a, s + e]);
		testOne(rexp, a + e + a, [a, e, a]);
		testOne(rexp, a + e + s, [a, e, s]);
		testOne(rexp, a + e + e, [a, e, e]);
		
		testOne(rexp, s + a + a, [s, a, a]);
		testOne(rexp, s + a + s, [s, a, s]);
		testOne(rexp, s + a + e, [s, a, e]);
		testOne(rexp, s + s + a, [s, s, a]);
		testOne(rexp, s + s + s, [s, s, s]);
		testOne(rexp, s + s + e, [s, s + e]);
		testOne(rexp, s + e + a, [s + e, a]);
		testOne(rexp, s + e + s, [s + e, s]);
		testOne(rexp, s + e + e, [s + e, e]);
		
		testOne(rexp, e + a + a, [e, a, a]);
		testOne(rexp, e + a + s, [e, a, s]);
		testOne(rexp, e + a + e, [e, a, e]);
		testOne(rexp, e + s + a, [e, s, a]);
		testOne(rexp, e + s + s, [e, s, s]);
		testOne(rexp, e + s + e, [e, s + e]);
		testOne(rexp, e + e + a, [e, e, a]);
		testOne(rexp, e + e + s, [e, e, s]);
		testOne(rexp, e + e + e, [e, e, e]);
	});
}



function testOne(rexp, s, expectedResult) {
	it('oneTest', function() {
		const a = s.match(rexp);
		assert.deepStrictEqual(a, expectedResult);
		assert.strictEqual(a.length, expectedResult.length);
	});
}
```

Native implementation and my implementation pass the test but babel implementation does not.

**Comment by nicolo-ribaudo** [comment]
@mathiasbynens This might be a bug with regexpu-core?

**Comment by mathiasbynens** [comment]
From <https://github.com/mathiasbynens/regenerate#regenerateprototypetostringoptions>:

> Note that lone low surrogates cannot be matched accurately using regular expressions in JavaScript. Regenerate’s output makes a best-effort approach but [there can be false negatives in this regard](https://github.com/mathiasbynens/regenerate/issues/28#issuecomment-72224808).

Lookbehind assertions make this possible but they’re not widely shipped yet so we just cannot rely on them for now.

**Comment by mathiasbynens** [comment]
@vitaliylag Do you want to submit a PR to Regenerate?

**Comment by vitaliylag** [comment]
I don't know what is Regenerate. Also the bug is with high surrogates but not with low surrogates. I just don't understand why the Babel process "." character such strange way. Right regular expression is so easy and also really makes sense.

**Comment by mathiasbynens** [comment]
Babel uses regexpu-core which uses Regenerate, which is the source of this bug. Would you be open to submitting a patch based on your suggested RegExp?

Your test uses U+DC00 which is a [low surrogate or trail surrogate](https://mathiasbynens.be/notes/javascript-encoding#surrogate-pairs).

**Comment by vitaliylag** [comment]
> Your test uses U+DC00 which is a low surrogate or trail surrogate.

Yes, I confused them :)

> Would you be open to submitting a patch based on your suggested RegExp?

I do not know how does Regenerate work so I do not know what can I change to fix this bug. Searching in the code does not match bad regexp so I guess I cannot create a PR.

**Comment by mathiasbynens** [comment]
I’m happy to give you some pointers! https://github.com/mathiasbynens/regenerate/blob/d34ae9be511306a872b742d8421ae18c4ffdc0bf/regenerate.js#L1020-L1036

**Comment by vitaliylag** [comment]
No, it looks too difficult to understand it all.

**Comment by mathiasbynens** [comment]
@vitaliylag Okay. The underlying problem is this one: https://gist.github.com/mathiasbynens/bbe7f870208abcfec860 Can you think of a regular expression that solves this problem without using lookbehind assertions? If so, please let me know, and I’ll happily make the change in Regenerate.

**Comment by mathiasbynens** [comment]
@vitaliylag Ok, here’s the underlying problem:

```js
const regex = /^a(?:[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])b$/;

console.assert(regex.test('a\uD834b') === true);
console.assert(regex.test('a\uDC00b') === true);
console.assert(regex.test('a\uD834\uDF06b') === false);

console.log('You did it!');
```

Can you think of a regular expression that passes all the assertions without using lookbehind assertions? If so, please let me know.

**Comment by jridgewell** [comment]
`/^a(?:[\uD800-\uDBFF](?![\uDC00-\uDFFF])|[\uDC00-\uDFFF])b$/`?

**Comment by vitaliylag** [comment]
What for do you need to match lone surrogates? What is the original regexp that we need to convert to ES5? If it was `/^a.b$/u` then third call should return true not false. Regexp I suggested already returns true so everything is good.

**Comment by vitaliylag** [comment]
I guess I know what the problem you want to solve. When I call something like this `var set = regenerate().add('\uD800')` you should find only lone surrogates. That is not related with dot but interesting problem.

**Comment by vitaliylag** [comment]
But should you even use regenerate to process dots? I guess you add all ranges (0-9, 11-12, 14-... and so on) but I don't know why you should add ranges because dot is special case and Regenerate is not needed for it.

**Comment by mathiasbynens** [comment]
@jridgewell Right, I should have done a better job of creating a test case. Let me try again:

```js
{
	// The goal is to match lone surrogates only.
	const regex = /([\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/;

	// Lone high surrogate U+D800:
	console.assert('\uD800'.match(regex)[1] === '\uD800');
	// Lone low surrogate U+DC00:
	console.assert('\uDC00'.match(regex)[1] === '\uDC00');
	// Surrogate pair (i.e. no lone surrogates):
	console.assert('\uD834\uDF06'.match(regex) === null);
}

{
	// This must be the same regular expression as above, with only `^a` added at
	// the beginning and `b$` added at the end.
	const regex = /^a([\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])b$/;

	// Lone high surrogate U+D800:
	console.assert('a\uD800b'.match(regex)[1] === '\uD800');
	// Lone low surrogate U+DC00:
	console.assert('a\uDC00b'.match(regex)[1] === '\uDC00');
	// Surrogate pair (i.e. no lone surrogates):
	console.assert('a\uD834\uDF06b'.match(regex) === null);
}

console.log('You did it!');
```

@vitaliylag `/./u` is just one example of a regexp that matches lone surrogates. There are infinitely more, and we cannot special-case them all. That’s why a generic solution is needed.

**Comment by mathiasbynens** [comment]
Edited my earlier comment to respond to some of @vitaliylag’s comments. Posting here for those reading email notifications:


@vitaliylag

> […] find only lone surrogates. That is not related with dot but interesting problem.

It is related because `/./u` matches lone surrogates.

> But should you even use regenerate to process dots? I guess you add all ranges (0-9, 11-12, 14-... and so on) but I don't know why you should add ranges because dot is special case and Regenerate is not needed for it.

`/./u` is just one example of a regexp that matches lone surrogates. There are infinitely more, and we cannot special-case them all. That’s why a generic solution is needed.

**Comment by vitaliylag** [comment]
Well.
1. You need match high surrogate. It is easy because of (?!).
2. You need match low surrogate. Actually it is easy too. This is because of current position cannot be at end of surrogate pair. So you just need check [\uDC00-\uDFFF]. That's all.

Of course you can ask why current position cannot be at end of surrogate pair.
- If you was searching range of code points >= 65536 then you wrote something this:
   ```js
   /( \uD877[\uDF55-\uDFFF]   |   [\uD878-\uD87A][\uDC00-\uDFFF]   |   \uD87B[\uDC00-\uDC01] )/
   
   //(but without spaces of course)
   ```
   It can match only full surrogate pairs. This example matches code points from 188245 to 191489. I have chosen this range randomly. Even if you will use *?+ you will match full surrogates anyway.
- If you was searching range of code points <= 55295 you cannot go to the end of surrogate pair.
- If you was searching high surrogate everything will be OK because of look ahead using (?!)
- If you was searching low surrogate you cannot be at the end of surrogate pair because end is already passed.

So as result you don't need look behind but you do need look ahead.

**Comment by mathiasbynens** [comment]
@vitaliylag Not sure I follow. Which regex would you use to solve [the latest example I posted](https://github.com/babel/babel/issues/6931#issuecomment-348069324)?

**Comment by vitaliylag** [comment]
I forgot third case:
3. You need match low surrogate but NOTHING is before low surrogate in current regexp (beginning of regexp). Example
   ```js
   /\uDC00 blablabla/ug
   ```
   I have know no solutions for this case. But this is very and very rare case so I guess you just should throw exception that this regexp cannot be converted to ES5. You don't need throw exception in Regenerator but in Babel itself.

> Which regex would you use to solve the latest example I posted?

This is easy.

---
**Example 1 (high surrogate):**

Original regexp:
```js
/abc \uD800 fgh/ug
```

Converted regexp:
```js
/abc \uD800(?![\uDC00-\uDFFF]) fgh/g
```

---
**Example 2 (low surrogate):**

Original regexp:
```js
/abc \uDC00 fgh/ug
```

Converted regexp:
```js
/abc \uDC00 fgh/g   //it works right in all cases because when we search \uDC00 current
                    //position cannot be equal position of some low surrogate that is
                    //end of surrogate pair.
```

---
**Example 3 (low surrogate at the start):**

```js
/\uDC00 fgh/ug

throw new BabelError('Syntax error');   //You cannot convert it to ES5
```

---
**Example 4 (low surrogate that looks like start but not):**

```js
/^\uDC00 fgh/ug
```
well be converted to:
```js
/\uDC00 fgh/g
```

---
**Example 5 (low surrogate that does not look like start but is a start):**

```js
/(a|b*)\uDC00 fgh/ug

throw new BabelError('Syntax error');   //You cannot convert it to ES5
```

---
**Example 6 (high surrogate):**

This example shows why you need look ahead check

Original regexp:
```js
/abc \uD800./ug
```

Converted regexp:
```js
/abc \uD800(?![\uDC00-\uDFFF])(HERE IS DOT EXPRESSION)/g
```

---
**Example 7 (high surrogate):**

Interesting example

Original regexp:
```js
/abc \uD800(\uDC00)/ug
```

Converted regexp:
```js
/abc \uD800(?![\uDC00-\uDFFF])(\uDC00)/g   //match nothing but that is right!
```

---
**Example 8 (high surrogate):**

Interesting example 2

Original regexp:
```js
/abc \uD800(\uDC00|d)/ug
```

Converted regexp:
```js
/abc \uD800(?![\uDC00-\uDFFF])(\uDC00|d)/g   //match only "abc \uD800d" and that is right
```

**Comment by vitaliylag** [comment]
Check lone high surrogate:
```js
/[\uD800-\uDBFF](?![\uDC00-\uDFFF])/
```

Check lone low surrogate:
```js
/[\uDC00-\uDFFF]/   //does not work at the start of regexp
```

Did you understand my solution and why you do not need lookbehind?

**Comment by vitaliylag** [comment]
I added fixed version of Babel to tests in first comment (it is called "babel-fixed").

But it'll be great if you optimize dot to my version because of performance and file size. Now dot is sum of:
1) usual chars
2) astral chars
3) lone high surrogates
4) lone low surrogates

but you can optimize it to:
1) usual chars + lone low surrogates
2) lone high surrogates + astral chars

**Comment by mathiasbynens** [comment]
> You need match low surrogate but NOTHING is before low surrogate in current regexp (beginning of regexp). Example

That’s an easy case as we can just use `^`. This is what the current Regenerate output does.

I’m still a little confused — could you please post the full regular expression that solves https://github.com/babel/babel/issues/6931#issuecomment-348069324?

**Comment by vitaliylag** [comment]
> That’s an easy case as we can just use ^. This is what the current Regenerate output does.

No, you cannnot do this.

Example:
```js
/\uDC00/u.test('aaa \uDC00');       //returns true
/\uDC00/u.test('aaa \uD800\uDC00'); //returns false
```
But if you try to convert it then:
```js
/^\uDC00/.test('aaa \uDC00');       //returns false (wrong result)
/^\uDC00/.test('aaa \uD800\uDC00'); //returns false
```

> This is what the current Regenerate output does.

No. It does this:
```js
(?:[^\uD800-\uDBFF]|^)
```
And it fails the tests because this version captures character it should not have.

It fails not only in tests with "match" but even in tests with calling "test" function:
```js
/a \uDC00/u.test('a \uDC00');        //return true
/a (?:[^\uD800-\uDBFF]|^)\uDC00/.test('a \uDC00'); //but after Babel returns false. This is bug
```

My solution has not such bugs (and any bugs at all). But yeah, my solution cannot work with lone high surrogate at the start of regexp but this is very rare case and I don't even think somebody needs it.

 

> I’m still a little confused — could you please post the full regular expression that solves #6931 (comment)?

To match all lone surrogates you just need match lone high and lone low surrogates:

```js
/[\uD800-\uDBFF](?![\uDC00-\uDFFF])|[\uDC00-\uDFFF]/
```

so the answer is:

```js
{
	// The goal is to match lone surrogates only.
	const regex = / ([\uD800-\uDBFF](?![\uDC00-\uDFFF])|[\uDC00-\uDFFF])/; //I added space because Babel should throw SyntaxError when matching lone low surrogate in the start of regexp
	
	// Lone high surrogate U+D800:
	console.assert(' \uD800'.match(regex)[1] === '\uD800');
	// Lone low surrogate U+DC00:
	console.assert(' \uDC00'.match(regex)[1] === '\uDC00');
	// Surrogate pair (i.e. no lone surrogates):
	console.assert(' \uD834\uDF06'.match(regex) === null);
}

{
	// This must be the same regular expression as above, with only `^a` added at
	// the beginning and `b$` added at the end.
	const regex = /^a([\uD800-\uDBFF](?![\uDC00-\uDFFF])|[\uDC00-\uDFFF])b$/;

	// Lone high surrogate U+D800:
	console.assert('a\uD800b'.match(regex)[1] === '\uD800');
	// Lone low surrogate U+DC00:
	console.assert('a\uDC00b'.match(regex)[1] === '\uDC00');
	// Surrogate pair (i.e. no lone surrogates):
	console.assert('a\uD834\uDF06b'.match(regex) === null);
}
```

Of course it passes all tests.

**Comment by vitaliylag** [comment]
You try to make "lookbehind" in your solution. But you don't need it: current char cannot be part of surrogate pair. So you just need to delete this "lookbehind" and everything will start working right.

**Comment by jridgewell** [comment]
@mathiasbynens: yours isn't solvable without look-behind. However, I'm not sure it has to be solved:

```js
var a = 'a',                        //usual character
    s = String.fromCharCode(55296), //start of surrogate pair
    e = String.fromCharCode(56320); //end of surrogate pair
r = /(?:[\0-\t\x0B\f\x0E-\u2027\u202A-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|[\uDC00-\uDFFF])/g;

console.log((a).match(r));
console.log((s).match(r));
console.log((e).match(r));
console.log((a + s).match(r));
console.log((a + e).match(r));
console.log((s + e).match(r));
console.log((s + s).match(r));
console.log((e + e).match(r));
```
